dgg-utilities.user.js

```diff
@@ -683,6 +684,172 @@ function injectScript() {
+    const match = TWITCH_EMBED_RE.exec(window.parent.location.hash);
+    return match ?
+      `https://www.twitch.tv/embed/${match[1]}/chat?parent=www.destiny.gg&darkpopout` :
+      null;
+  }
+
+  function activateEmbedChat(embedChatURL) {
+    embedChatActive = true;
+
+    // only update the src attribute if it has changed to avoid unnecessary refresh
+    if (embedChatIFrame.getAttribute("src") !== embedChatURL) {
+      embedChatIFrame.setAttribute("src", embedChatURL);
+    }
+
+    dggChatIFrame.style.display = "none";
+    embedChatIFrame.style.display = "block";
+
+    embedChatToggle.innerHTML = dggChatToggleLabel;
+  }
+
+  function deactivateEmbedChat() {
+    embedChatActive = false;
+
+    embedChatIFrame.style.display = "none";
+    dggChatIFrame.style.display = "block";
+
+    embedChatToggle.innerHTML = isLive() ? ytChatToggleLabel : embedChatToggleLabel;
+  }
+
+  function toggleEmbedChat() {
+    if (embedChatActive) {
+      deactivateEmbedChat();
+      return;
+    }
+
+    const ytEmbedChatURL = getYTEmbedChatURL();
+    if (ytEmbedChatURL) {
+      activateEmbedChat(ytEmbedChatURL);
+      return;
+    }
+
+    const twitchEmbedChatURL = getTwitchEmbedChatURL();
+    if (twitchEmbedChatURL) {
+      activateEmbedChat(twitchEmbedChatURL);
+      return;
+    }
+
+    const ytLiveChatURL = getYTLiveChatURL();
+    if (ytLiveChatURL) {
+      activateEmbedChat(ytLiveChatURL);
+      return;
+    }
+  }
+
+  function addEmbedChatToggleBtn() {
+    if (livePill == undefined) {
+      return;
+    }
+
+    if (window.parent.document.getElementById("embed-chat-iframe")) {
+      removeEmbedChatToggleBtn();
+    }
+
+    // add the iframe for the embedded chat
+    embedChatIFrame = document.createElement("iframe");
+    embedChatIFrame.id = "embed-chat-iframe"
+    embedChatIFrame.style.display = "none";
+    embedChatIFrame.setAttribute("seamless", "seamless");
+    dggChatIFrame.parentNode.appendChild(embedChatIFrame);
+
+    // add the link/button for toggling the embedded chat
+    embedChatToggle = document.createElement("a");
+    embedChatToggle.id = "embed-chat-toggle";
+    embedChatToggle.className = "float-left";
+    embedChatToggle.style.width = "100px";
+    embedChatToggle.innerHTML = isLive() ? ytChatToggleLabel : embedChatToggleLabel;
+
+    embedChatToggle.addEventListener("click", toggleEmbedChat);
+
+    window.parent.document.getElementById("chat-panel-tools").insertBefore(
+      embedChatToggle,
+      window.parent.document.getElementById("refresh").nextSibling
+    );
+
+    // The height styling of the DGG chat's input element will be messed up if the DGG chat isn't visible when the chat's iframe is refreshed
+    // https://github.com/destinygg/chat-gui/blob/78910027663171870a314cc3ab3c066334b72326/assets/chat/js/chat.js#L889
+    // So, show the DGG chat before refreshing the iframe
+    window.parent.document.getElementById("refresh").addEventListener("click", deactivateEmbedChat);
```

Comment:
This line is to cover a weird edge case where refreshing the chat iframe while the DGG chat is hidden causes it to load incorrectly. So I catch the refresh button being pressed and switch to the DGG chat quickly before refreshing.